driver-core/src/main/com/mongodb/client/model/expressions/Expression.java

```diff
@@ -118,4 +120,17 @@ public interface Expression {
      * @return
      */
     StringExpression asString();
+
+    /**
+     * Applies the given function to this argument. Note that "apply" usually
+     * applies functions to arguments; here, the parameters are reversed.
+     *
+     * @param f
+     * @return
+     * @param <T>
+     * @param <R>
+     */
+    <T extends Expression, R extends Expression> R apply(Function<T, R> f);
```

Comment:
[Noticed this function in https://github.com/mongodb/mongo-java-driver/pull/1059 and decided to leave a comment here]

The problem with this function is that when called on `IntegerExpression` it passes `Expression` into `f`, i.e., it erases the information about the type. Let's introduce `IntegerExpression.applyToInteger`/`applyToBoolean`/etc. methods?

Reply:
Discussed. These will have to be renamed. We considered various alternative names for `apply`: `numberApply` and `numberPassTo` are unclear, while `applyToNumber` sounds backwards. We came to `numberInto`, which has a good form, and finally agreed on `passNumberTo`, which puts the operation "pass" at the start. The length of the method is not a problem because it should not be used often, and should be noticable when it is used. Incidentally, this resolves my prior issue with using the term "apply" (the opposite of "apply" is "pass"). For switch, we'll follow the same pattern `switchNumber` and so on.

To re-iterate, the problem is that the type `T` above is not `TSelf`, and has nothing to do with the actual type that the `apply` method is invoked upon. So in the example above, an empty string is effectively cast into an integer, and then used in division. A similar problem arises with `Array<T>`, but there we at least have `T`, so we can use the `v -> v` trick for reductive operations. The approach of copying a method down is similar to what was done for add/multiply/etc. for numbers and integers, except that in this case, the method names cannot be the same due to type erasure.